File: gget-java/src/main/java/cz/zakjan/gget/visitor/GetByPathVisitor.java
package cz.zakjan.gget.visitor;
import cz.zakjan.gget.parser.ObjectPathBaseVisitor;
import cz.zakjan.gget.parser.ObjectPathParser;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

public class GetByPathVisitor extends ObjectPathBaseVisitor<Object> {
    private Object data;

    public GetByPathVisitor(Object data) {
        super();
        this.data = data;
    }

    @Override
    public Object visitStart(ObjectPathParser.StartContext ctx) {
        if (ctx.expression() != null) {
            return this.evaluateExpression(this.data, ctx.expression());
        }
        return this.data;
    }

    private Object evaluateExpression(Object data, ObjectPathParser.ExpressionContext ctx) {
        if (ctx instanceof ObjectPathParser.ParenthesesContext) {
            return this.evaluateParentheses(data, (ObjectPathParser.ParenthesesContext)ctx);
        } else if (ctx instanceof ObjectPathParser.RootObjectContext) {
            return this.evaluateRootObject(data, (ObjectPathParser.RootObjectContext)ctx);
        } else if (ctx instanceof ObjectPathParser.CurrentObjectContext) {
            return this.evaluateCurrentObject(data, (ObjectPathParser.CurrentObjectContext)ctx);
        } else if (ctx instanceof ObjectPathParser.PropertyAccessContext) {
            return this.evaluatePropertyAccess(data, (ObjectPathParser.PropertyAccessContext)ctx);
        } else if (ctx instanceof ObjectPathParser.IndexAccessContext) {
            return this.evaluateIndexAccess(data, (ObjectPathParser.IndexAccessContext)ctx);
        } else if (ctx instanceof ObjectPathParser.ArrayFilterContext) {
            return this.evaluateArrayFilter(data, (ObjectPathParser.ArrayFilterContext)ctx);
        } else if (ctx instanceof ObjectPathParser.ArrayMapContext) {
            return this.evaluateArrayMap(data, (ObjectPathParser.ArrayMapContext)ctx);
        } else if (ctx instanceof ObjectPathParser.FunctionContext) {
            return this.evaluateFunction(data, (ObjectPathParser.FunctionContext)ctx);
        } else if (ctx instanceof ObjectPathParser.UnaryContext) {
            return this.evaluateUnary(data, (ObjectPathParser.UnaryContext)ctx);
        } else if (ctx instanceof ObjectPathParser.UnaryLogicalNotContext) {
            return this.evaluateUnaryLogicalNot(data, (ObjectPathParser.UnaryLogicalNotContext)ctx);
        } else if (ctx instanceof ObjectPathParser.MultiplicativeContext) {
            return this.evaluateMultiplicative(data, (ObjectPathParser.MultiplicativeContext)ctx);
        } else if (ctx instanceof ObjectPathParser.AdditiveContext) {
            return this.evaluateAdditive(data, (ObjectPathParser.AdditiveContext)ctx);
        } else if (ctx instanceof ObjectPathParser.RelationalContext) {
            return this.evaluateRelational(data, (ObjectPathParser.RelationalContext)ctx);
        } else if (ctx instanceof ObjectPathParser.EqualityContext) {
            return this.evaluateEquality(data, (ObjectPathParser.EqualityContext)ctx);
        } else if (ctx instanceof ObjectPathParser.LogicalAndContext) {
            return this.evaluateLogicalAnd(data, (ObjectPathParser.LogicalAndContext)ctx);
        } else if (ctx instanceof ObjectPathParser.LogicalOrContext) {
            return this.evaluateLogicalOr(data, (ObjectPathParser.LogicalOrContext)ctx);
        } else if (ctx instanceof ObjectPathParser.TernaryContext) {
            return this.evaluateTernary(data, (ObjectPathParser.TernaryContext)ctx);
        } else if (ctx instanceof ObjectPathParser.StringContext) {
            return this.evaluateString(data, (ObjectPathParser.StringContext)ctx);
        } else if (ctx instanceof ObjectPathParser.NumberContext) {
            return this.evaluateNumber(data, (ObjectPathParser.NumberContext)ctx);
        } else if (ctx instanceof ObjectPathParser.BooleanContext) {
            return this.evaluateBoolean(data, (ObjectPathParser.BooleanContext)ctx);
        } else if (ctx instanceof ObjectPathParser.NullContext) {
            return this.evaluateNull(data, (ObjectPathParser.NullContext)ctx);
        } else {
            throw new IllegalStateException();
        }
    }

    private Object evaluateParentheses(Object data, ObjectPathParser.ParenthesesContext ctx) {
        return this.evaluateExpression(data, ctx.expression());
    }

    private Object evaluateRootObject(Object data, ObjectPathParser.RootObjectContext ctx) {
        return this.data;
    }

    private Object evaluateCurrentObject(Object data, ObjectPathParser.CurrentObjectContext ctx) {
        return data;
    }

    private Object evaluatePropertyAccess(Object data, ObjectPathParser.PropertyAccessContext ctx) {
        Object value = ctx.expr1 != null ? this.evaluateExpression(data, ctx.expr1) : data;
        String property = ctx.IDENTIFIER().getText();

        if (value instanceof Map && ((Map<String, Object>)value).containsKey(property)) {
            return ((Map<String, Object>)value).get(property);
        }
        return null;
    }

    private Object evaluateIndexAccess(Object data, ObjectPathParser.IndexAccessContext ctx) {
        Object value = ctx.expr1 != null ? this.evaluateExpression(data, ctx.expr1) : data;
        Object index = this.evaluateExpression(data, ctx.expr2);

        if (value instanceof Map && index instanceof String && ((Map<String, Object>)value).containsKey(index)) {
            return ((Map<String, Object>)value).get(index);
        } else if (value instanceof List && index instanceof Integer) {
            Integer normalizedIndex = (Integer)index + ((Integer)index < 0 ? ((List<Object>)value).size() : 0);
            if (normalizedIndex >= 0 && normalizedIndex < ((List<Object>)value).size()) {
                return ((List<Object>)value).get(normalizedIndex);
            }
        }
        return null;
    }

    private Object evaluateArrayFilter(Object data, ObjectPathParser.ArrayFilterContext ctx) {
        Object value = ctx.expr1 != null ? this.evaluateExpression(data, ctx.expr1) : data;

        if (value instanceof List) {
            return ((List<Object>)value).stream().filter(item -> {
                return this.toBoolean(this.evaluateExpression(item, ctx.expr2));
            }).collect(Collectors.toList());
        }
        return null;
    }

    private Object evaluateArrayMap(Object data, ObjectPathParser.ArrayMapContext ctx) {
        Object value = ctx.expr1 != null ? this.evaluateExpression(data, ctx.expr1) : data;

        if (value instanceof List) {
            return ((List<Object>)value).stream().map(item -> {
                return this.evaluateExpression(item, ctx.expr2);
            }).collect(Collectors.toList());
        }
        return null;
    }

    private Object evaluateFunction(Object data, ObjectPathParser.FunctionContext ctx) {
        String identifier = ctx.IDENTIFIER().getText();
        List<Object> args = ctx.expression().stream().map(ctx2 -> {
            return this.evaluateExpression(data, ctx2);
        }).collect(Collectors.toList());

        return GetByPathVisitorFunctions.callFunction(identifier, args);
    }

    private Object evaluateUnary(Object data, ObjectPathParser.UnaryContext ctx) {
        Object value = this.evaluateExpression(data, ctx.expression());

        if (value instanceof Integer) {
            if (ctx.op.getText().equals("+")) {
                return +(Integer)value;
            } else if (ctx.op.getText().equals("-")) {
                return -(Integer)value;
            } else {
                throw new IllegalStateException();
            }
        }
        return null;
    }

    private Object evaluateUnaryLogicalNot(Object data, ObjectPathParser.UnaryLogicalNotContext ctx) {
        Object value = this.evaluateExpression(data, ctx.expression());

        return !this.toBoolean(value);
    }

    private Object evaluateMultiplicative(Object data, ObjectPathParser.MultiplicativeContext ctx) {
        Object left = this.evaluateExpression(data, ctx.expression().get(0));
        Object right = this.evaluateExpression(data, ctx.expression().get(1));

        if (left instanceof Integer && right instanceof Integer) {
            if (ctx.op.getText().equals("*")) {
                return (Integer)left * (Integer)right;
            } else if (ctx.op.getText().equals("/")) {
                if (!right.equals(0)) {
                    return (Integer)left / (Integer)right;
                } else {
                    return null;
                }
            } else {
                throw new IllegalStateException();
            }
        }
        return null;
    }

    private Object evaluateAdditive(Object data, ObjectPathParser.AdditiveContext ctx) {
        Object left = this.evaluateExpression(data, ctx.expression().get(0));
        Object right = this.evaluateExpression(data, ctx.expression().get(1));

        if (left instanceof String && right instanceof String) {
            if (ctx.op.getText().equals("+")) {
                return (String)left + (String)right;
            } else {
                throw new IllegalStateException();
            }
        } else if (left instanceof Integer && right instanceof Integer) {
            if (ctx.op.getText().equals("+")) {
                return (Integer)left + (Integer)right;
            } else if (ctx.op.getText().equals("-")) {
                return (Integer)left - (Integer)right;
            } else {
                throw new IllegalStateException();
            }
        }
        return null;
    }

    private Object evaluateRelational(Object data, ObjectPathParser.RelationalContext ctx) {
        Object left = this.evaluateExpression(data, ctx.expression().get(0));
        Object right = this.evaluateExpression(data, ctx.expression().get(1));

        if (left instanceof Integer && right instanceof Integer) {
            if (ctx.op.getText().equals("<")) {
                return (Integer)left < (Integer)right;
            } else if (ctx.op.getText().equals(">")) {
                return (Integer)left > (Integer)right;
            } else if (ctx.op.getText().equals("<=")) {
                return (Integer)left <= (Integer)right;
            } else if (ctx.op.getText().equals(">=")) {
                return (Integer)left >= (Integer)right;
            } else {
                throw new IllegalStateException();
            }
        }
        return null;
    }

    private Object evaluateEquality(Object data, ObjectPathParser.EqualityContext ctx) {
        Object left = this.evaluateExpression(data, ctx.expression().get(0));
        Object right = this.evaluateExpression(data, ctx.expression().get(1));

        if (ctx.op.getText().equals("==")) {
            return Objects.equals(left, right);
        } else if (ctx.op.getText().equals("!=")) {
            return !Objects.equals(left, right);
        } else {
            throw new IllegalStateException();
        }
    }

    private Object evaluateLogicalAnd(Object data, ObjectPathParser.LogicalAndContext ctx) {
        Object left = this.evaluateExpression(data, ctx.expression().get(0));
        Object right = this.evaluateExpression(data, ctx.expression().get(1));

        return this.toBoolean(left) == true ? right : left;
    }

    private Object evaluateLogicalOr(Object data, ObjectPathParser.LogicalOrContext ctx) {
        Object left = this.evaluateExpression(data, ctx.expression().get(0));
        Object right = this.evaluateExpression(data, ctx.expression().get(1));

        return this.toBoolean(left) == true ? left : right;
    }

    private Object evaluateTernary(Object data, ObjectPathParser.TernaryContext ctx) {
        Object condition = this.evaluateExpression(data, ctx.expression().get(0));
        Object pass = this.evaluateExpression(data, ctx.expression().get(1));
        Object fail = this.evaluateExpression(data, ctx.expression().get(2));

        return this.toBoolean(condition) == true ? pass : fail;
    }

    private Object evaluateString(Object data, ObjectPathParser.StringContext ctx) {
        return ctx.STRING().getText().substring(1, ctx.STRING().getText().length() - 1);
    }

    private Object evaluateNumber(Object data, ObjectPathParser.NumberContext ctx) {
        try {
            return Integer.parseInt(ctx.NUMBER().getText());
        } catch (NumberFormatException e) {
            return Long.parseLong(ctx.NUMBER().getText());
        }
    }

    private Object evaluateBoolean(Object data, ObjectPathParser.BooleanContext ctx) {
        return ctx.BOOLEAN().getText().equals("true");
    }

    private Object evaluateNull(Object data, ObjectPathParser.NullContext ctx) {
        return null;
    }

    private Boolean toBoolean(Object data) {
        return data instanceof Boolean ? (Boolean)data == true : data != null;
    }
}

File: objectpath-ts/src/visitor/GetByPathVisitor.ts
import { ObjectPathBaseVisitor } from "../parser/ObjectPathBaseVisitor";
import * as ObjectPathParser from "../parser/ObjectPathParser";
import { GetByPathVisitorFunctions } from "./GetByPathVisitorFunctions";

export class GetByPathVisitor extends ObjectPathBaseVisitor<unknown> {
    constructor(private data: unknown) {
        super();
    }

    visitStart(ctx: ObjectPathParser.StartContext): unknown {
        const expression = ctx.expression();
        if (expression) {
            return this.evaluateExpression(this.data, expression);
        }
        return this.data;
    }

    private evaluateExpression(data: unknown, ctx: ObjectPathParser.ExpressionContext): unknown {
        if (ctx instanceof ObjectPathParser.ParenthesesContext) {
            return this.evaluateParentheses(data, ctx);
        } else if (ctx instanceof ObjectPathParser.RootObjectContext) {
            return this.evaluateRootObject(data, ctx);
        } else if (ctx instanceof ObjectPathParser.CurrentObjectContext) {
            return this.evaluateCurrentObject(data, ctx);
        } else if (ctx instanceof ObjectPathParser.DotAccessContext) {
            return this.evaluateDotAccess(data, ctx);
        } else if (ctx instanceof ObjectPathParser.BracketAccessContext) {
            return this.evaluateBracketAccess(data, ctx);
        } else if (ctx instanceof ObjectPathParser.ArrayFilterContext) {
            return this.evaluateArrayFilter(data, ctx);
        } else if (ctx instanceof ObjectPathParser.ArrayMapContext) {
            return this.evaluateArrayMap(data, ctx);
        } else if (ctx instanceof ObjectPathParser.FunctionContext) {
            return this.evaluateFunction(data, ctx);
        } else if (ctx instanceof ObjectPathParser.UnaryContext) {
            return this.evaluateUnary(data, ctx);
        } else if (ctx instanceof ObjectPathParser.UnaryLogicalNotContext) {
            return this.evaluateUnaryLogicalNot(data, ctx);
        } else if (ctx instanceof ObjectPathParser.MultiplicativeContext) {
            return this.evaluateMultiplicative(data, ctx);
        } else if (ctx instanceof ObjectPathParser.AdditiveContext) {
            return this.evaluateAdditive(data, ctx);
        } else if (ctx instanceof ObjectPathParser.RelationalContext) {
            return this.evaluateRelational(data, ctx);
        } else if (ctx instanceof ObjectPathParser.EqualityContext) {
            return this.evaluateEquality(data, ctx);
        } else if (ctx instanceof ObjectPathParser.NullishCoalescingContext) {
            return this.evaluateNullishCoalescing(data, ctx);
        } else if (ctx instanceof ObjectPathParser.LogicalAndContext) {
            return this.evaluateLogicalAnd(data, ctx);
        } else if (ctx instanceof ObjectPathParser.LogicalOrContext) {
            return this.evaluateLogicalOr(data, ctx);
        } else if (ctx instanceof ObjectPathParser.ConditionalContext) {
            return this.evaluateConditional(data, ctx);
        } else if (ctx instanceof ObjectPathParser.StringContext) {
            return this.evaluateString(data, ctx);
        } else if (ctx instanceof ObjectPathParser.NumberContext) {
            return this.evaluateNumber(data, ctx);
        } else if (ctx instanceof ObjectPathParser.BooleanContext) {
            return this.evaluateBoolean(data, ctx);
        } else if (ctx instanceof ObjectPathParser.NullContext) {
            return this.evaluateNull(data, ctx);
        } else {
            /* istanbul ignore next */
            throw new Error("Illegal state");
        }
    }

    private evaluateParentheses(data: unknown, ctx: ObjectPathParser.ParenthesesContext): unknown {
        return this.evaluateExpression(data, ctx.expression());
    }

    private evaluateRootObject(_data: unknown, _ctx: ObjectPathParser.RootObjectContext): unknown {
        return this.data;
    }

    private evaluateCurrentObject(data: unknown, _ctx: ObjectPathParser.CurrentObjectContext): unknown {
        return data;
    }

    private evaluateDotAccess(data: unknown, ctx: ObjectPathParser.DotAccessContext): unknown {
        const value = ctx._expr1 ? this.evaluateExpression(data, ctx._expr1) : data;
        const property = ctx.IDENTIFIER().text;

        if (this.isObject(value) && value.hasOwnProperty(property)) {
            return value[property];
        }
        return null;
    }

    private evaluateBracketAccess(data: unknown, ctx: ObjectPathParser.BracketAccessContext): unknown {
        const value = ctx._expr1 ? this.evaluateExpression(data, ctx._expr1) : data;
        const index = this.evaluateExpression(data, ctx._expr2);

        if (this.isObject(value) && typeof index === "string" && value.hasOwnProperty(index)) {
            return value[index];
        } else if (Array.isArray(value) && typeof index === "number") {
            const normalizedIndex = index + (index < 0 ? value.length : 0);
            if (normalizedIndex >= 0 && normalizedIndex < value.length) {
                return value[normalizedIndex];
            }
        }
        return null;
    }

    private evaluateArrayFilter(data: unknown, ctx: ObjectPathParser.ArrayFilterContext): unknown {
        const value = ctx._expr1 ? this.evaluateExpression(data, ctx._expr1) : data;

        if (Array.isArray(value)) {
            return value.filter(item => {
                return this.toBoolean(this.evaluateExpression(item, ctx._expr2));
            });
        }
        return null;
    }

    private evaluateArrayMap(data: unknown, ctx: ObjectPathParser.ArrayMapContext): unknown {
        const value = ctx._expr1 ? this.evaluateExpression(data, ctx._expr1) : data;

        if (Array.isArray(value)) {
            return value.map(item => {
                return this.evaluateExpression(item, ctx._expr2);
            });
        }
        return null;
    }

    private evaluateUnary(data: unknown, ctx: ObjectPathParser.UnaryContext): unknown {
        const value = this.evaluateExpression(data, ctx.expression());

        if (typeof value === "number") {
            if (ctx._op.text === "+") {
                return +value;
            } else if (ctx._op.text === "-") {
                return -value;
            } else {
                /* istanbul ignore next */
                throw new Error("Illegal state");
            }
        }
        return null;
    }

    private evaluateUnaryLogicalNot(data: unknown, ctx: ObjectPathParser.UnaryLogicalNotContext): unknown {
        const value = this.evaluateExpression(data, ctx.expression());

        return !this.toBoolean(value);
    }

    private evaluateMultiplicative(data: unknown, ctx: ObjectPathParser.MultiplicativeContext): unknown {
        const left = this.evaluateExpression(data, ctx.expression()[0]);
        const right = this.evaluateExpression(data, ctx.expression()[1]);

        if (typeof left === "number" && typeof right === "number") {
            if (ctx._op.text === "*") {
                return left * right;
            } else if (ctx._op.text === "/") {
                if (right !== 0) {
                    return Math.trunc(left / right);
                } else {
                    return null;
                }
            } else {
                /* istanbul ignore next */
                throw new Error("Illegal state");
            }
        }
        return null;
    }

    private evaluateAdditive(data: unknown, ctx: ObjectPathParser.AdditiveContext): unknown {
        const left = this.evaluateExpression(data, ctx.expression()[0]);
        const right = this.evaluateExpression(data, ctx.expression()[1]);

        if (typeof left === "string" && typeof right === "string") {
            if (ctx._op.text === "+") {
                return left + right;
            } else {
                /* istanbul ignore next */
                throw new Error("Illegal state");
            }
        } else if (typeof left === "number" && typeof right === "number") {
            if (ctx._op.text === "+") {
                return left + right;
            } else if (ctx._op.text === "-") {
                return left - right;
            } else {
                /* istanbul ignore next */
                throw new Error("Illegal state");
            }
        }
        return null;
    }

    private evaluateRelational(data: unknown, ctx: ObjectPathParser.RelationalContext): unknown {
        const left = this.evaluateExpression(data, ctx.expression()[0]);
        const right = this.evaluateExpression(data, ctx.expression()[1]);

        if (typeof left === "number" && typeof right === "number") {
            if (ctx._op.text === "<") {
                return left < right;
            } else if (ctx._op.text === ">") {
                return left > right;
            } else if (ctx._op.text === "<=") {
                return left <= right;
            } else if (ctx._op.text === ">=") {
                return left >= right;
            } else {
                /* istanbul ignore next */
                throw new Error("Illegal state");
            }
        }
        return null;
    }

    private evaluateEquality(data: unknown, ctx: ObjectPathParser.EqualityContext): unknown {
        const left = this.evaluateExpression(data, ctx.expression()[0]);
        const right = this.evaluateExpression(data, ctx.expression()[1]);

        if (ctx._op.text === "==") {
            return left === right;
        } else if (ctx._op.text === "!=") {
            return left !== right;
        } else {
            /* istanbul ignore next */
            throw new Error("Illegal state");
        }
    }

    private evaluateNullishCoalescing(data: unknown, ctx: ObjectPathParser.NullishCoalescingContext): unknown {
        const left = this.evaluateExpression(data, ctx.expression()[0]);
        const right = this.evaluateExpression(data, ctx.expression()[1]);

        return left === null ? right : left;
    }

    private evaluateLogicalAnd(data: unknown, ctx: ObjectPathParser.LogicalAndContext): unknown {
        const left = this.evaluateExpression(data, ctx.expression()[0]);
        const right = this.evaluateExpression(data, ctx.expression()[1]);

        return this.toBoolean(left) === true ? right : left;
    }

    private evaluateLogicalOr(data: unknown, ctx: ObjectPathParser.LogicalOrContext): unknown {
        const left = this.evaluateExpression(data, ctx.expression()[0]);
        const right = this.evaluateExpression(data, ctx.expression()[1]);

        return this.toBoolean(left) === true ? left : right;
    }

    private evaluateConditional(data: unknown, ctx: ObjectPathParser.ConditionalContext): unknown {
        const condition = this.evaluateExpression(data, ctx.expression()[0]);
        const pass = this.evaluateExpression(data, ctx.expression()[1]);
        const fail = this.evaluateExpression(data, ctx.expression()[2]);

        return this.toBoolean(condition) === true ? pass : fail;
    }

    private evaluateFunction(data: unknown, ctx: ObjectPathParser.FunctionContext): unknown {
        const identifier = ctx.IDENTIFIER().text;
        const args = ctx.expression().map(ctx2 => {
            return this.evaluateExpression(data, ctx2);
        });

        return GetByPathVisitorFunctions.callFunction(identifier, args);
    }

    private evaluateString(_data: unknown, ctx: ObjectPathParser.StringContext): unknown {
        return ctx.STRING().text.substring(1, ctx.STRING().text.length - 1);
    }

    private evaluateNumber(_data: unknown, ctx: ObjectPathParser.NumberContext): unknown {
        return parseInt(ctx.NUMBER().text, 10);
    }

    private evaluateBoolean(_data: unknown, ctx: ObjectPathParser.BooleanContext): unknown {
        return ctx.BOOLEAN().text === "true";
    }

    private evaluateNull(_data: unknown, _ctx: ObjectPathParser.NullContext): unknown {
        return null;
    }

    private toBoolean(data: unknown): boolean {
        return typeof data === "boolean" ? data === true : data != null;
    }

    private isObject(data: unknown): data is { [key: string]: unknown } {
        return data && typeof data === 'object';
    }
}
